packages/progress-monitor/src/browser/progress-dialog.ts

```diff
@@ -0,0 +1,92 @@
+/********************************************************************************
+ * Copyright (C) 2018 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject, postConstruct } from 'inversify';
+import { Message } from '@phosphor/messaging';
+import { Disposable } from '@theia/core';
+import { Widget, BaseWidget, Key } from '@theia/core/lib/browser';
+import { ProgressWidget } from './progress-widget';
+
+@injectable()
+export class ProgressDialog extends BaseWidget {
+
+    protected contentNode: HTMLDivElement;
+
+    protected isOpen: boolean = false;
+
+    @inject(ProgressWidget) readonly widget: ProgressWidget;
+
+    constructor() {
+        super();
+        this.contentNode = document.createElement('div');
+        this.contentNode.classList.add(ProgressWidget.Styles.PROGRESS_MONITOR_CONTENT);
+
+        this.node.classList.add(ProgressWidget.Styles.PROGRESS_WIDGET_DIALOG);
+
+        this.node.appendChild(this.contentNode);
+        this.update();
+    }
+
+    @postConstruct()
+    protected init() {
+        this.toDispose.push(this.widget);
+    }
+
+    protected onUpdateRequest(msg: Message): void {
+        super.onUpdateRequest(msg);
+        this.widget.update();
+    }
+
+    protected onAfterAttach(msg: Message): void {
+        super.onAfterAttach(msg);
+        Widget.attach(this.widget, this.contentNode);
+        this.toDisposeOnDetach.push(Disposable.create(() =>
+            Widget.detach(this.widget)
+        ));
+
+        this.addKeyListener(document.body, Key.ESCAPE, e => {
+            this.closeDialog();
+        });
+        this.addEventListener(document.body, 'click', e => {
```

Comment:
How about using a real callback: https://github.com/theia-ide/theia/blob/80f402c621fe197130eae7abd22a0d89008b4ef1/packages/core/src/browser/status-bar/status-bar.tsx#L45

Reply:
The onclick for the status bar is set [here](https://github.com/theia-ide/theia/blob/0725529c0df47d81e0306064aae6f59caf68331e/packages/progress-monitor/src/browser/progress-status-bar.ts#L56) to toggle whether the modal window is open. That part is needed so that it closes the dialog if anything but the modal window is clicked.

Reply:
sorry, I don't understand then why code checks that a clicked element should not belong to status bar entry. Should not it check that it does not belong to `this.host`?

Reply:
I was trying to make it so that you could click anywhere and close the modal window. However, that ran into some problems because when you would click the status bar item it would first hit that piece of code and close the dialog then the status bar item onclick would be triggered resulting in an opened dialog. To get around this I just made sure that if the status bar is the item being clicked then it shouldn't close the window here and instead wait for the onclick of the status bar item to close the dialog.

Reply:
ok

You can get rid of couplign with the status bar entry by making sure that your event listener is first, either by using the capture pahse or not attaching a dialog to a status bar entry, since it has higher z-index.

Also you should use `target` not `srcElement`: https://developer.mozilla.org/en-US/docs/Web/API/Event/srcElement